### ISSUE TITLE
`.git` suffix inconsistently enforced

### ISSUE THREAD

**Issue by willstott101** [gabrielcsapo/node-git-server#101]
When creating a new repo, a `.git` suffix is appended here: https://github.com/gabrielcsapo/node-git-server/blob/e25c481b85f8b929146b4ea46227d0b6fd83caf4/src/git.ts#L288 and it appears to be enforced in `list`, but not everywhere

However the `.git` suffix is not a requirement anywhere else - this server is perfectly happy serving git repositories without a `.git` suffix afaict... https://github.com/gabrielcsapo/node-git-server/blob/e25c481b85f8b929146b4ea46227d0b6fd83caf4/src/util.ts#L150

This inconsistency has tripped me up a bit. I would personally be in favour of enforcing a .git suffix everywhere - possibly by redirecting to `.git` suffixed URL if a non-git suffixed URL is encountered.

This is inconsistent with bare/checked-out repo conventions though... which might suggest to enforce `.git` only if `checkout` is `true`...

I think for us we're going to enforce and stick to the suffix externally for the meantime - but it'd be good to gather opinions of other maintainers/users.

**Comment by gabrielcsapo** [comment]
I agree, we could make this an option that could be "enforceGitSuffix"? Thoughts on what it would look like for opt-in behavior?

**Comment by batchor** [comment]
This may be fixed in our fork ([@batchfy/node-git-server](https://github.com/batchfy/node-git-server)) — see https://github.com/batchfy/node-git-server/pull/1, in case it's useful here.

The approach taken there: every request resolves to one canonical repo name up front, which is then used for the dir resolver, `exists()`, `create()` and the emitted events.

- Canonical name is `<name>.git` for bare repos and `<name>` when `checkout` is set, so the checked-out convention is respected (option 3 in the issue).
- A repo that already exists under the requested name wins, so repos stored without the suffix keep working, and both URL forms reach the same repository.
- `list()` now reports directories that actually are git repositories instead of filtering on the `.git` suffix.
- `enforceGitSuffix: true` (per @gabrielcsapo's suggestion above) makes `<name>.git` the only name a repo is known by: GET requests for a suffix-less URL get a 301 to the `.git` URL, and non-suffixed directories are neither listed nor served.

Worth noting the pre-existing bugs this surfaced: because `create()` appended `.git` while nothing else did, `exists()` returned false for an existing repo (404 with `autoCreate: false`, and a redundant `git init` on every request otherwise), and `GET /:repo/HEAD` 404'd for a repo that clones fine. It only looked like it worked because git itself resolves `repos/doom` to `repos/doom.git`.
